Issues with wildcard usage in paths containing names starting with digits

Hello,

In our project, we have encountered an issue with folder names in our DAM that include numbers. When we specify ACL rules targeting a specific path directly, everything functions as expected. However, an error arises if we attempt to use a wildcard in the path specification.

For instance, consider the following rule:

```
- path: '/content/dam/brands/21c/global-marketing/*/'
   permission: allow
   privileges: jcr:read
```
This setup results in an error stating: `"Could not process yaml files / e=javax.jcr.query.InvalidQueryException: java.text.ParseException: Query: /jcr:root/content/dam/brands/"21c"()/global-marketing/; expected: jcr:root, /, *, @, (, ."`

It's important to note that if we use the path /content/dam/brands/21c/global-marketing without the wildcard, it operates correctly.

@Vocinglero Thanks, I could reproduce. The full stack trace from the exception is as follows:

```
javax.jcr.query.InvalidQueryException: java.text.ParseException: Query:_/jcr:root/content/dam/brands/21(*)c/global-marketing/*; expected: jcr:root, /, *, @, (, .
	at org.apache.jackrabbit.oak.jcr.query.QueryManagerImpl.executeQuery(QueryManagerImpl.java:149) [org.apache.jackrabbit.oak-jcr:1.60.0.T20240131102219-0cde853]
	at org.apache.jackrabbit.oak.jcr.query.QueryImpl$2.perform(QueryImpl.java:105) [org.apache.jackrabbit.oak-jcr:1.60.0.T20240131102219-0cde853]
	at org.apache.jackrabbit.oak.jcr.query.QueryImpl$2.perform(QueryImpl.java:101) [org.apache.jackrabbit.oak-jcr:1.60.0.T20240131102219-0cde853]
	at org.apache.jackrabbit.oak.jcr.delegate.SessionDelegate.perform(SessionDelegate.java:229) [org.apache.jackrabbit.oak-jcr:1.60.0.T20240131102219-0cde853]
	at org.apache.jackrabbit.oak.jcr.query.QueryImpl.execute(QueryImpl.java:100) [org.apache.jackrabbit.oak-jcr:1.60.0.T20240131102219-0cde853]
	at biz.netcentric.cq.tools.actool.helper.QueryHelper.getNodesFromQuery(QueryHelper.java:154)
	at biz.netcentric.cq.tools.actool.helper.QueryHelper.getNodePathsFromQuery(QueryHelper.java:164)
	at biz.netcentric.cq.tools.actool.helper.QueryHelper.getNodePathsFromQuery(QueryHelper.java:132)
	at biz.netcentric.cq.tools.actool.configreader.YamlConfigReader.handleWildcards(YamlConfigReader.java:297)
	at biz.netcentric.cq.tools.actool.configreader.YamlConfigReader.getPreservedOrderdAceSet(YamlConfigReader.java:280)
	at biz.netcentric.cq.tools.actool.configreader.YamlConfigReader.getAceConfigurationBeans(YamlConfigReader.java:126)
	at biz.netcentric.cq.tools.actool.configreader.YamlConfigurationMerger.getMergedConfigurations(YamlConfigurationMerger.java:188)
	at biz.netcentric.cq.tools.actool.impl.AcInstallationServiceImpl.installConfigurationFiles(AcInstallationServiceImpl.java:286)
	at biz.netcentric.cq.tools.actool.impl.AcInstallationServiceImpl.apply(AcInstallationServiceImpl.java:217)
...
Caused by: java.text.ParseException: Query:_/jcr:root/content/dam/brands/21(*)c/global-marketing/*; expected: jcr:root, /, *, @, (, .
	at org.apache.jackrabbit.oak.query.xpath.XPathToSQL2Converter.getSyntaxError(XPathToSQL2Converter.java:1222) [org.apache.jackrabbit.oak-core:1.60.0.T20240131102219-0cde853]
	at org.apache.jackrabbit.oak.query.xpath.XPathToSQL2Converter.getSyntaxError(XPathToSQL2Converter.java:1212) [org.apache.jackrabbit.oak-core:1.60.0.T20240131102219-0cde853]
	at org.apache.jackrabbit.oak.query.xpath.XPathToSQL2Converter.convertToStatement(XPathToSQL2Converter.java:364) [org.apache.jackrabbit.oak-core:1.60.0.T20240131102219-0cde853]
	at org.apache.jackrabbit.oak.query.xpath.XPathToSQL2Converter.convert(XPathToSQL2Converter.java:99) [org.apache.jackrabbit.oak-core:1.60.0.T20240131102219-0cde853]
	at org.apache.jackrabbit.oak.query.QueryEngineImpl.parseQuery(QueryEngineImpl.java:194) [org.apache.jackrabbit.oak-core:1.60.0.T20240131102219-0cde853]
	at org.apache.jackrabbit.oak.query.QueryEngineImpl.executeQuery(QueryEngineImpl.java:267) [org.apache.jackrabbit.oak-core:1.60.0.T20240131102219-0cde853]
	at org.apache.jackrabbit.oak.jcr.query.QueryManagerImpl.executeQuery(QueryManagerImpl.java:137) [org.apache.jackrabbit.oak-jcr:1.60.0.T20240131102219-0cde853]
	... 178 common frames omitted
```

--- a/accesscontroltool-bundle/src/main/java/biz/netcentric/cq/tools/actool/configreader/YamlConfigReader.java
+++ b/accesscontroltool-bundle/src/main/java/biz/netcentric/cq/tools/actool/configreader/YamlConfigReader.java
@@ -27,6 +27,7 @@ import javax.jcr.Session;
 import javax.jcr.query.InvalidQueryException;
 
 import org.apache.commons.lang3.StringUtils;
+import org.apache.jackrabbit.util.ISO9075;
 import org.apache.sling.jcr.api.SlingRepository;
 import org.osgi.service.component.annotations.Component;
 import org.osgi.service.component.annotations.Reference;
@@ -293,7 +294,7 @@ public class YamlConfigReader implements ConfigReader {
             final AceBean tmpAclBean) throws InvalidQueryException,
             RepositoryException {
         // perform query using the path containing wildcards
-        final String query = "/jcr:root" + tmpAclBean.getJcrPath();
+        final String query = createXPathQueryForPathWithWildcards(tmpAclBean.getJcrPath());
         final Set<String> result = QueryHelper.getNodePathsFromQuery(session, query);
 
         if (result.isEmpty()) {
@@ -314,6 +315,19 @@ public class YamlConfigReader implements ConfigReader {
                 }
             }
         }
+    }
+
+    /**
+     * Create an XPath query for the given JCR path (which may contain wildcards ({@code *}) which should be preserved).
+     * @param jcrPathWithWildcard a valid JCR path where some names/elements might be wildcards.
+     * 
+     * @see <a href="https://jackrabbit.apache.org/archive/wiki/JCR/EncodingAndEscaping_115513396.html#EncodingAndEscaping-Encodingpathinqueries">Encoding Path In XPath Queries</a>
+     * @see <a href="https://issues.apache.org/jira/browse/JCR-5051">JCR-5051</a>
+     */
+    static final String createXPathQueryForPathWithWildcards(String jcrPathWithWildcard) {
+        final String query = "/jcr:root" + ISO9075.encodePath(jcrPathWithWildcard);
+        // decode wildcards again (https://issues.apache.org/jira/browse/JCR-5051)
+        return query.replace("_x002a_", "*");
     }
 
     protected AceBean getNewAceBean() {

--- a/accesscontroltool-bundle/src/test/java/biz/netcentric/cq/tools/actool/configreader/YamlConfigReaderTest.java
+++ b/accesscontroltool-bundle/src/test/java/biz/netcentric/cq/tools/actool/configreader/YamlConfigReaderTest.java
@@ -154,6 +154,12 @@ public class YamlConfigReaderTest {
                 () -> yamlConfigReader.getUserConfigurationBeans(yamlList, null));
     }
 
+    @Test
+    public void testCreateXPathQueryForPathWithWildcards() {
+        assertEquals("/jcr:root/var/test", YamlConfigReader.createXPathQueryForPathWithWildcards("/var/test"));
+        assertEquals("/jcr:root/var/*/_x0031_2node/*", YamlConfigReader.createXPathQueryForPathWithWildcards("/var/*/12node/*"));
+    }
+
     static List<Map> getYamlList(final String filename) throws IOException {
         final String configString = getTestConfigAsString(filename);
 

